### PR TITLE
Show a warning when PDF does not have selectable text

### PR DESCRIPTION
**Pull request by robertknight**
This is an initial solution for https://github.com/hypothesis/product-backlog/issues/1081. It adds a banner to the top of the PDF viewer when the current PDF does not contain any selectable text.

Current design (updated 2020-10-16):

pdf-no-text-v2

The "OCR-optimized" text links to https://web.hypothes.is/help/how-to-ocr-optimize-pdfs/.

### CODE REVIEW

File: src/annotator/anchoring/pdf.js [hunk 1]
```diff
@@ -79,16 +79,32 @@ async function getPageView(pageIndex) {
     // a "pdfPage" property.
     pageView = await new Promise(resolve => {
       const onPagesLoaded = () => {
-        document.removeEventListener('pagesloaded', onPagesLoaded);
+        pdfViewer.eventBus.off('pagesloaded', onPagesLoaded);
         resolve(pdfViewer.getPageView(pageIndex));
       };
-      document.addEventListener('pagesloaded', onPagesLoaded);
+      pdfViewer.eventBus.on('pagesloaded', onPagesLoaded);
```
Comment (robertknight) [inline src/annotator/anchoring/pdf.js:85]: This fixes a compatibility issue with new versions of PDF.js that we hadn't noticed before. PDF viewer events are no longer emitted as DOM events. It should be extracted into a separate PR and shipped first.

**Comment by lyzadanger** [comment]
@robertknight I've pushed some commits that apply styling to the warning message. As of 152212e it looks like this:

image

The border width (borders are top and bottom only) is `2px` currently. We can go to `3px`, but it looks a little chunky.

File: src/annotator/plugin/pdf.js [hunk 2]
```diff
@@ -50,6 +90,69 @@ export default class PDF extends Delegator {
...
+      await this.uri();
+    } catch (e) {
+      return;
+    }
+
+    try {
+      const hasText = await pdfAnchoring.documentHasText();
+      this._showNoSelectableTextWarning(!hasText);
+    } catch (err) {
+      /* istanbul ignore next */
```
Comment (lyzadanger) [inline src/annotator/plugin/pdf.js:108]: I suppose I could have used this approach to what we were discussing regarding #2646 and the guarding against arrays being of different lengths...

File: src/annotator/plugin/pdf.js [hunk 3]
```diff
@@ -50,6 +90,69 @@ export default class PDF extends Delegator {
...
+   */
+  _showNoSelectableTextWarning(showWarning) {
+    if (!showWarning) {
+      this._warningBanner?.remove();
+      this._warningBanner = null;
+      return;
+    }
+
+    const mainContainer = /** @type {HTMLElement} */ (document.querySelector(
+      '#mainContainer'
```
Comment (lyzadanger) [inline src/annotator/plugin/pdf.js:127]: Is this OK as a literal here (i.e. is this selector used elsewhere and could it be maintained as a constant somewhere?)

File: src/annotator/plugin/pdf.js [hunk 4]
```diff
@@ -50,6 +90,69 @@ export default class PDF extends Delegator {
...
+    ));
+
+    const updateBannerHeight = () => {
+      /* istanbul ignore next */
+      if (!this._warningBanner) {
+        return;
+      }
+      const rect = this._warningBanner.getBoundingClientRect();
```
Comment (lyzadanger) [inline src/annotator/plugin/pdf.js:135]: A comment here might not hurt—explaining that we're moving the main container down to accommodate the warning banner, right?

File: src/annotator/plugin/pdf.js [hunk 5]
```diff
@@ -50,6 +90,69 @@ export default class PDF extends Delegator {
...
+      mainContainer.style.top = rect.height + 'px';
+      this._warningBanner.style.top = -rect.height + 'px';
+    };
+
+    this._warningBanner = document.createElement('div');
+    this._warningBanner.className =
```
Comment (lyzadanger) [inline src/annotator/plugin/pdf.js:141]: I'll acknowledge that these classnames could be...tidier.

File: src/annotator/plugin/pdf.js [hunk 6]
```diff
@@ -50,6 +90,69 @@ export default class PDF extends Delegator {
...
+      'annotator-pdf-warning-banner annotator-pdf-warning-banner--notice';
+    mainContainer.appendChild(this._warningBanner);
+    render(<WarningBanner />, this._warningBanner);
+
+    // @ts-expect-error - TS is missing `ResizeObserver`
+    if (typeof ResizeObserver !== 'undefined') {
+      // Update the banner when the window is resized or the Hypothesis
+      // sidebar is opened.
+      // @ts-ignore
+      new ResizeObserver(updateBannerHeight).observe(this._warningBanner);
```
Comment (lyzadanger) [inline src/annotator/plugin/pdf.js:151]: `ResizeObserver` is supported in, it looks like, all of our target browsers save for Android's Firefox. Historically, we've used `MutationObserver` (broader/older support but more complex API) as a polyfill, see `src/sidebar/util/observe-element-size` — do we have a plan going forward?

**Review by lyzadanger — APPROVED** [review]
Looking good.

File: src/annotator/plugin/pdf.js (hunk 3 above)
Comment (robertknight) [reply src/annotator/plugin/pdf.js:127]: This selector isn't referenced anywhere else. It could do with a comment explaining where the element comes from though.

File: src/annotator/plugin/pdf.js (hunk 6 above)
Comment (robertknight) [reply src/annotator/plugin/pdf.js:158]: `ResizeObserver` was only recently introduced in Safari (13.1) which means it isn't available on older iOS devices. In older browsers the banner height will simply be static based on the window size when the client first loads. I think that's an acceptable failure mode.
